File: mqttclients/dotnet/MQTTnet.Client.Extensions/CommandServer.cs
using MQTTnet.Protocol;
using System.Text;

namespace MQTTnet.Client.Extensions;

public abstract class CommandServer<T, TResp>
{
    private readonly IMqttClient _mqttClient;
    private readonly string _commandName;

    public Func<T, Task<TResp>>? OnCommandReceived { get; set; }

    private readonly string _requestTopic = string.Empty;

    private readonly IMessageSerializer _serializer;

    public CommandServer(IMqttClient mqttClient, string cmdName, IMessageSerializer serializer)
    {
        _mqttClient = mqttClient;
        _commandName = cmdName;
        _serializer = serializer;

        RequestTopicAttribute? rta = GetType().GetCustomAttributes(true).OfType<RequestTopicAttribute>().FirstOrDefault();
        _requestTopic = rta!.Topic.Replace("{clientId}", _mqttClient.Options.ClientId).Replace("{commandName}", _commandName);

        mqttClient.ApplicationMessageReceivedAsync += async m =>
        {
            string topic = m.ApplicationMessage.Topic;
            if (topic.Equals(_requestTopic))
            {
                if (m.ApplicationMessage.ContentType != serializer.ContentType)
                {
                    throw new ApplicationException($"Invalid content type. Expected :{_serializer.ContentType} Actual :{m.ApplicationMessage.ContentType}");
                }

                T request = _serializer.FromBytes<T>(m.ApplicationMessage.PayloadSegment.Array!);

                try
                {
                    TResp response = await OnCommandReceived?.Invoke(request)!;

                    byte[] respBytes = _serializer.ToBytes(response);

                    MqttClientPublishResult pubAck = await mqttClient.PublishAsync(new MqttApplicationMessageBuilder()
                        .WithTopic(m.ApplicationMessage.ResponseTopic)
                        .WithContentType(_serializer.ContentType)
                        .WithPayload(respBytes)
                        .WithUserProperty("Status", 200.ToString())
                        .WithCorrelationData(m.ApplicationMessage.CorrelationData)
                        .Build());
                }
                catch (Exception ex)
                {
                    MqttClientPublishResult pubAck = await mqttClient.PublishAsync(new MqttApplicationMessageBuilder()
                        .WithTopic(m.ApplicationMessage.ResponseTopic)
                        .WithCorrelationData(m.ApplicationMessage.CorrelationData)
                        .WithPayload(Encoding.UTF8.GetBytes(ex.Message))
                        .WithUserProperty("status", 500.ToString())
                        .Build());
                }
            }
        };
    }

    public async Task StartAsync(CancellationToken ct = default) =>
        await _mqttClient.SubscribeAsync(_requestTopic, MqttQualityOfServiceLevel.AtLeastOnce, ct);
}

File: mqttclients/ts/mqttjsClientExtensions/src/commandServer.ts
import {
    IPublishPacket,
    MqttClient
} from 'mqtt';
import {
    logger,
    IMessageSerializer
} from './';

const ModuleName = 'CommandServer';

export class CommandServer<T, TResp>
{
    private mqttClient: MqttClient;
    private requestTopic = '';
    private commandName: string;
    private serializer: IMessageSerializer;
    public onCommandReceived: (request: T) => TResp;

    constructor(mqttClient: MqttClient, requestTopicPattern: string, commandName: string, serializer: IMessageSerializer) {
        this.mqttClient = mqttClient;
        this.commandName = commandName;
        this.serializer = serializer;

        this.requestTopic = requestTopicPattern.replace('{clientId}', mqttClient.options.clientId ?? '').replace('{commandName}', this.commandName);

        this.mqttClient.on('message', async (topic: string, payload: Buffer, packet: IPublishPacket) => {
            if (topic === this.requestTopic) {
                if (packet.properties?.contentType !== serializer.contentType) {
                    throw new Error(`Invalid content type. Expected :${this.serializer.contentType} Actual :${packet.properties?.contentType}`);
                }

                const request = this.serializer.fromBytes<T>(payload);
                const responseTopic = packet.properties.responseTopic ?? '';
                const requestCorrelationId = packet.properties?.correlationData;

                try {
                    logger.info({ tags: [ModuleName] }, `Received command request on topic: ${this.requestTopic}, with correlationId: ${requestCorrelationId?.toString()}`);

                    const response = this.onCommandReceived(request);
                    const respBytes = this.serializer.toBytes(response);

                    const pubAck = await this.mqttClient.publishAsync(responseTopic, respBytes, {
                        qos: 1,
                        properties: {
                            contentType: this.serializer.contentType,
                            correlationData: packet.properties?.correlationData,
                            userProperties: {
                                status: '200'
                            }
                        }
                    });

                    logger.info({ tags: [ModuleName] }, `Published success response with mid: ${pubAck?.messageId} on topic: ${responseTopic}, with correlationId: ${requestCorrelationId?.toString()}`);
                }
                catch (ex) {
                    const pubAck = await this.mqttClient.publishAsync(responseTopic, Buffer.from(ex.message), {
                        qos: 1,
                        properties: {
                            correlationData: packet.properties?.correlationData,
                            userProperties: {
                                status: '500'
                            }
                        }
                    });

                    logger.warn({ tags: [ModuleName] }, `Published error response with mid: ${pubAck?.messageId} on topic: ${responseTopic}`);
                }
            }
        });
    }

    public async startAsync(): Promise<void> {
        await this.mqttClient.subscribeAsync(this.requestTopic, {
            qos: 1
        });
    }
}